### DISCUSSION TITLE
`@types/ember__component` 4.0.6 update breaks `Helper.compute` without signature

### DISCUSSION THREAD

**Discussion by mydea** [typed-ember/ember-cli-typescript#1498] [Q&A] [answered]
The following code for helpers worked fine so far:

```ts
import Helper from '@ember/component/helper';

export default class MyHelper extends Helper {
  compute([name]: [string]) {
    return name;
  }
}
```

Now throws a type error, as there is no signature defined.  I guess that means I have to duplicate the type declaration in the signature for the `compute` method now? It is a bit unfortunate that it cannot be inferred, as  for example with something like this:

```ts
type PropName = 'prop1' | 'prop2';

interface HelperSignature {
  PositionalArgs: [obj: MyModel, propName: PropName];
}

export default class MyHelper extends Helper<HelperSignature> {
  compute([obj, propName]: [MyModel, PropName]) {
    // do something
  }
}
```

It becomes somewhat cumbersome to type this twice. Am I missing something here, or is this the only way to type a helper class like this going forward?

**Comment by chriskrycho** [comment, marked as answer]
Good question! I didn't actually realize folks were using that pattern, but it makes sense. This is basically the tradeoff of making class-based helpers natively support [Glint](https://github.com/typed-ember/glint), and is an unfortunate limitation of the Typescript type system as it stands: methods whose values are related to the generic type of the class they are members of don't get inference based on the generic types. This is partly to allow subclassing to work as folks expect, and partly just because it's a hard thing TS hasn't solved yet.

I've turned this into a discussion and am leaving it here so folks can share suggestions, alternative designs, etc.!

One option here is:

```ts
interface Args {
  Positional: [obj: MyModel, propName: 'prop1' | 'prop2'];
}

export default class MyHelper extends Helper<{ Args: Args }> {
  compute([obj, propName]: Args['Positional']) {
    // do something
  }
}
```

That has the overhead of a little bit of repetition but is definitely much shorter.

**Reply by chriskrycho** [reply]
Update (and related to @jamescdavis's comment below: this should be resolved on the latest version of `@types/ember__component` (courtesy of DefinitelyTyped/DefinitelyTyped#59715): Glint will error if you use this form, but it will continue to type check in a non-Glint application.

(Heads up: it is likely that we will *remove* support for this form when integrating natively into Ember's own types, precisely because the value proposition is highest when you’re using Glint, and we want to avoid leading people down unhappy paths—so this unblocks you for today, but you still still plan to migrate to the `Signature`-based form!)

**Comment by jamescdavis** [comment]
Here's a possible solution: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59657
